docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+ - [PaxosCommit](https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/PaxosCommit.tla)
+
+
+More advanced code patterns over variants can be found in [Raft][]:
+
+```tla
+Receive(m) ==
+    LET i == m.mdest
+        j == m.msource
+    IN \* Any RPC with a newer term causes the recipient to advance
+       \* its term first. Responses with stale terms are ignored.
+       \/ UpdateTerm(i, j, m)
+       \/ /\ m.mtype = RequestVoteRequest
+          /\ HandleRequestVoteRequest(i, j, m)
+       \/ /\ m.mtype = RequestVoteResponse
+          /\ \/ DropStaleResponse(i, j, m)
+             \/ HandleRequestVoteResponse(i, j, m)
+       \/ /\ m.mtype = AppendEntriesRequest
+          /\ HandleAppendEntriesRequest(i, j, m)
+       \/ /\ m.mtype = AppendEntriesResponse
+          /\ \/ DropStaleResponse(i, j, m)
+             \/ HandleAppendEntriesResponse(i, j, m)
+    ...
+Next ==
+    ...
+           \/ \E m \in DOMAIN messages : Receive(m)
+           \/ \E m \in DOMAIN messages : DuplicateMessage(m)
+           \/ \E m \in DOMAIN messages : DropMessage(m)
+    ...
+```
+
+### 3. Other issues
+
+As can be seen from the Paxos example, we should take care of sets of
+records that are used as sets of variants:
+
+```tla
+Message ==      [type : {"1a"}, bal : Ballot]
+           \cup [type : {"1b"}, acc : Acceptor, bal : Ballot, 
+                 mbal : Ballot \cup {-1}, mval : Value \cup {None}]
+           \cup [type : {"2a"}, bal : Ballot, val : Value]
+           \cup [type : {"2b"}, acc : Acceptor, bal : Ballot, val : Value]
+```           
+
+## 4. Options
+
+There are several solutions to the issue of precise record typing.
+
+### 4.1. Support only plain records
+
+In this case, we would only allow to mix records that have exactly the same
+shape. As a result, when we use variants, we would have to add extra
+fields to all of them. This solution is not very different from the current
+implementation of record type checking, though it would allow us to quickly
+detect spelling errors.
+
+**Blocker.** This does not look like a real solution, as it would immediately
+render the existing examples invalid. Moreover, they would be no obvious way to
+repair these examples.
+
+### 4.2. Support plain records and variants, but no row typing
+
+In this scenario, the type checker would issue an error, if a record expression
+accesses a field that is outside of its type:
+
+```tla
+FieldAccess ==
+  LET m == [ a |-> 2, b |-> "B" ] IN
+  /\ m.a > 1        \* type OK
+  /\ m.b = "B"      \* type OK
+  /\ m.c = { 1, 2 } \* should flag a type error
+```
+
+Moreover, the following example would require a type annotation:
+
+```tla
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+In the above example, the type checker would not be able to infer the type of
+`m` and would require an explicit type annotation:
```

Comment:
1. We require type annotations for constants and variables. In some cases, the users have to write type annotations, when the type checker cannot decide between a function and a sequence. In most of the cases, we actually do not have to annotate user-defined operators. I see row types as a way to avoid unnecessary type annotations.

2. We do not have type declarations as in OCaml. The annotation `@typeAlias` is just an alias that is getting replaced with the type before the type checker starts doing anything. I am not even sure how OCaml does it.